### PR TITLE
Add support for the GHC JavaScript backend (node.js)

### PR DESCRIPTION
**Pull request by luite**
This PR adds support to the `process` package to target the `node.js` JavaScript platform using the GHC JavaScript backend.

This allows among other things the GHC JavaScript backend to use custom Cabal `Setup.hs` scripts.

### CODE REVIEW

File: jsbits/process.js [hunk 1]
```diff
@@ -0,0 +1,585 @@
+//#OPTIONS: CPP
+// XXX do we need this?
```
[the remark this answers is not in the document]
Comment (luite) [reply jsbits/process.js:2]: Testsuite still passes if removed

**Comment by bgamari** [comment]
What is the plan to ensure that this is tested once merged?

**Comment by luite** [comment]
The tests for `process` in the GHC testsuite can be enabled again for JS (except the ones that rely on extra C source files), they're currently marked broken.

`cabal test` works after `cabal configure "--with-compiler=javascript-unknown-ghcjs-ghc" "--with-hc-pkg=javascript-unknown-ghcjs-ghc-pkg" --enable-tests`, but I'm not sure if we can get such a  compiler in the GitHub CI matrix yet.

**Comment by bgamari** [comment]
> The tests for `process` in the GHC testsuite can be enabled again for JS (except the ones that rely on extra C source files), they're currently marked broken.

Yes, but in this case I'm a bit confused: At least some of these tests live in this repository and yet are still marked as broken. Are these really broken? Is there a list of tests which are fixed by this MR?

**Comment by doyougnu** [comment]
Ben is right, these tests should be enabled with this PR:

```
process003
process004
process006
process007
process008
process009
process010
process011
T1780
T4889
T8343
T4198
```

Ideally we would add a JS backend test runner to this repo so we can tests these tests. Unfortunately it looks like the CI for this repo uses `matrix` and `stack` to run the testsuite. So until we have stack support and an official release (I guess?) it'll be hard to test here unless we pulled in `haskell.nix` or set up a job using `ghcup` or something like that.

Perhaps we:
1) @luite enable these tests for js for this PR
2) open an MR on GHC that simply bumps the process and check the tests there

**Comment by luite** [comment]
Sorry for late reply, I'm on holiday this week. I'll add a commit remove the `js_broken` for tests that are now passing.

**Comment by hsyl20** [comment]
@luite Is there a corresponding GHC MR that uses this code?

**Comment by luite** [comment]
> @luite Is there a corresponding GHC MR that uses this code?

Ah I suppose it would be useful to get a CI run there. I'll make an MR

**Comment by luite** [comment]
GHC MR here: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10801

**Comment by luite** [comment]
I missed a few expected results in the `process` package, and also some `base` and `rts` tests pass now, which I've adjusted in MR !10801. After this has been merged, the `process` submodule bump should include these changes.

**Review by hsyl20 — APPROVED** [review]

**Pull request merged**